app/src/main/java/me/devcexx/call2alarm/MainActivity.kt

```diff
@@ -144,24 +165,50 @@ class MainActivity : ComponentActivity() {
         }
     }
 
-    private fun ensureServiceInCorrectState() {
-        GlobalScope.launch(Dispatchers.Main) {
-            logi(Thread.currentThread().toString())
-            val intent = Intent(baseContext, CallInterceptorService::class.java)
-            if (interceptionEnabled.value) {
-                if (permissionRequester.requestPermission(Manifest.permission.READ_PHONE_STATE) &&
-                    permissionRequester.requestPermission(Manifest.permission.ACCESS_NOTIFICATION_POLICY)) {
-                    permissionRequester.requestPermission(Manifest.permission.POST_NOTIFICATIONS) // Optional permission
-                    startForegroundService(intent)
-                } else {
-                    //interceptionEnabledState.value = false
-                    interceptionEnabled.value = false
-                    stopService(intent)
-                }
-            } else {
-                stopService(intent)
+
+    @Composable
+    private fun CheckForPermissions(permissionsRequester: MultiplePermissionsState) {
+        val context = LocalContext.current
+
+
+        val mandatoryPermissionsGranted by remember {
+            derivedStateOf {
+                permissionsRequester.checkMandatoryPermissions()
             }
         }
 
+        LaunchedEffect(mandatoryPermissionsGranted, context, interceptionEnabled.value, block = {
+            if (!mandatoryPermissionsGranted) {
+                permissionsRequester.launchMultiplePermissionRequest()
+            }
+            context.onPermissionsProvided(mandatoryPermissionsGranted && interceptionEnabled.value)
+        })
+    }
+
+    private fun MultiplePermissionsState.checkMandatoryPermissions(): Boolean {
+        val permissionMap =
+            this.permissions.associateBy { it.permission }
+        return mandatoryPermissions.all { permission ->
+            permissionMap[permission]?.status?.isGranted ?: false
+        }
+    }
+
+}
+
+
+private val mandatoryPermissions =
+    listOf(Manifest.permission.READ_PHONE_STATE, Manifest.permission.ACCESS_NOTIFICATION_POLICY)
+
+private val optionalPermissions = arrayOf(Manifest.permission.POST_NOTIFICATIONS)
```

Comment:
[nit] is there a reason for mandatoryPermissions to be defined as List, and keep this one as an array?